### PR TITLE
Return getBoundingVolume by value to work around HIP optimizer

### LINKED ISSUE #553: Why does HIP compiler produce slow assembly for a reasonable code?
In #552 we observed an interesting phenomenon: a simple change
```diff
--- a/src/details/ArborX_DetailsHappyTreeFriends.hpp
+++ b/src/details/ArborX_DetailsHappyTreeFriends.hpp
@@ -52,7 +52,7 @@ struct HappyTreeFriends
   }

   template <class BVH>
-  static KOKKOS_FUNCTION auto const &getBoundingVolume(BVH const &bvh, int i)
+  static KOKKOS_FUNCTION auto getBoundingVolume(BVH const &bvh, int i)
   {
     return bvh._internal_and_leaf_nodes(i).bounding_volume;
   }
```
resulted in a much much faster code on Spock. 

The results were observed with the 4.3 rocm compiler. The flag `-fno-legacy-pass-manager` did not make any difference.

The assembly code for the radius search traversal function (approximately) looked like this:
### Return by reference variant
```assembly
B32_5:                                 ;   Parent Loop BB32_3 Depth=1
                                        ; =>  This Loop Header: Depth=2
                                        ;       Child Loop BB32_23 Depth 3
                                        ;       Child Loop BB32_29 Depth 3
                                        ;       Child Loop BB32_34 Depth 3
                                        ;         Child Loop BB32_37 Depth 4
.Ltmp7516:
        .loc    57 57 12 is_stmt 1              ; src/details/ArborX_Point.hpp:57:12
        v_lshlrev_b64 v[15:16], 5, v[13:14]
        v_add_co_u32_e32 v18, vcc, v11, v15
        v_addc_co_u32_e32 v19, vcc, v12, v16, vcc
.Ltmp7517:
        .loc    183 99 20                       ; src/details/ArborX_DetailsAlgorithms.hpp:99:20
        flat_load_dword v15, v[18:19] offset:8
.Ltmp7518:
        .loc    57 57 12                        ; src/details/ArborX_Point.hpp:57:12
        v_add_co_u32_e32 v16, vcc, 8, v18
        v_addc_co_u32_e32 v17, vcc, 0, v19, vcc
.Ltmp7519:
        .loc    183 99 18                       ; src/details/ArborX_DetailsAlgorithms.hpp:99:18
        s_waitcnt vmcnt(0) lgkmcnt(0)
        v_cmp_nlt_f32_e32 vcc, v8, v15
        s_and_saveexec_b64 s[10:11], vcc
        s_cbranch_execz BB32_9
; %bb.6:                                ;   in Loop: Header=BB32_5 Depth=2
.Ltmp7520:
        .loc    57 57 12                        ; src/details/ArborX_Point.hpp:57:12
        v_lshlrev_b64 v[18:19], 5, v[13:14]
        v_add_co_u32_e32 v18, vcc, v11, v18
        v_addc_co_u32_e32 v19, vcc, v12, v19, vcc
.Ltmp7521:
        .loc    183 101 25                      ; src/details/ArborX_DetailsAlgorithms.hpp:101:25
        flat_load_dword v15, v[18:19] offset:20
        .loc    183 101 23 is_stmt 0            ; src/details/ArborX_DetailsAlgorithms.hpp:101:23
        s_waitcnt vmcnt(0) lgkmcnt(0)
        v_cmp_ngt_f32_e32 vcc, v8, v15
        s_and_saveexec_b64 s[12:13], vcc
.Ltmp7522:
; %bb.7:                                ;   in Loop: Header=BB32_5 Depth=2
        .loc    183 0 23                        ; src/details/ArborX_DetailsAlgorithms.hpp:0:23
        v_mov_b32_e32 v15, v8
; %bb.8:                                ; %Flow54
                                        ;   in Loop: Header=BB32_5 Depth=2
        s_or_b64 exec, exec, s[12:13]
BB32_9:                                 ;   in Loop: Header=BB32_5 Depth=2
        s_or_b64 exec, exec, s[10:11]
        .loc    183 99 20 is_stmt 1             ; src/details/ArborX_DetailsAlgorithms.hpp:99:20
        flat_load_dword v18, v[16:17] offset:4
        .loc    183 99 18 is_stmt 0             ; src/details/ArborX_DetailsAlgorithms.hpp:99:18
        s_waitcnt vmcnt(0) lgkmcnt(0)
        v_cmp_nlt_f32_e32 vcc, v9, v18
        s_and_saveexec_b64 s[10:11], vcc
        s_cbranch_execz BB32_13
; %bb.10:                               ;   in Loop: Header=BB32_5 Depth=2
.Ltmp7523:
        .loc    57 57 12 is_stmt 1              ; src/details/ArborX_Point.hpp:57:12
        v_lshlrev_b64 v[18:19], 5, v[13:14]
        v_add_co_u32_e32 v18, vcc, v11, v18
        v_addc_co_u32_e32 v19, vcc, v12, v19, vcc
.Ltmp7524:
        .loc    183 101 25                      ; src/details/ArborX_DetailsAlgorithms.hpp:101:25
        flat_load_dword v19, v[18:19] offset:24
        v_mov_b32_e32 v18, v9
        .loc    183 101 23 is_stmt 0            ; src/details/ArborX_DetailsAlgorithms.hpp:101:23
        s_waitcnt vmcnt(0) lgkmcnt(0)
        v_cmp_gt_f32_e32 vcc, v9, v19
        s_and_saveexec_b64 s[12:13], vcc
.Ltmp7525:
; %bb.11:                               ;   in Loop: Header=BB32_5 Depth=2
        .loc    183 0 23                        ; src/details/ArborX_DetailsAlgorithms.hpp:0:23
        v_mov_b32_e32 v18, v19
; %bb.12:                               ; %Flow51
                                        ;   in Loop: Header=BB32_5 Depth=2
        s_or_b64 exec, exec, s[12:13]
BB32_13:                                ;   in Loop: Header=BB32_5 Depth=2
        s_or_b64 exec, exec, s[10:11]
        .loc    183 99 20 is_stmt 1             ; src/details/ArborX_DetailsAlgorithms.hpp:99:20
        flat_load_dword v16, v[16:17] offset:8
        .loc    183 99 18 is_stmt 0             ; src/details/ArborX_DetailsAlgorithms.hpp:99:18
        s_waitcnt vmcnt(0) lgkmcnt(0)
        v_cmp_nlt_f32_e32 vcc, v10, v16
        s_and_saveexec_b64 s[10:11], vcc
        s_cbranch_execz BB32_17
; %bb.14:                               ;   in Loop: Header=BB32_5 Depth=2
.Ltmp7526:
        .loc    57 57 12 is_stmt 1              ; src/details/ArborX_Point.hpp:57:12
        v_lshlrev_b64 v[16:17], 5, v[13:14]
        v_add_co_u32_e32 v16, vcc, v11, v16
        v_addc_co_u32_e32 v17, vcc, v12, v17, vcc
.Ltmp7527:
        .loc    183 101 25                      ; src/details/ArborX_DetailsAlgorithms.hpp:101:25
        flat_load_dword v17, v[16:17] offset:28
        v_mov_b32_e32 v16, v10
        .loc    183 101 23 is_stmt 0            ; src/details/ArborX_DetailsAlgorithms.hpp:101:23
        s_waitcnt vmcnt(0) lgkmcnt(0)
        v_cmp_gt_f32_e32 vcc, v10, v17
        s_and_saveexec_b64 s[12:13], vcc
.Ltmp7528:
; %bb.15:                               ;   in Loop: Header=BB32_5 Depth=2
        .loc    183 0 23                        ; src/details/ArborX_DetailsAlgorithms.hpp:0:23
        v_mov_b32_e32 v16, v17
; %bb.16:                               ; %Flow49
                                        ;   in Loop: Header=BB32_5 Depth=2
```
### Return by value variant
```assembly
BB32_5:                                 ;   Parent Loop BB32_3 Depth=1
                                        ; =>  This Loop Header: Depth=2
                                        ;       Child Loop BB32_23 Depth 3
                                        ;       Child Loop BB32_29 Depth 3
                                        ;       Child Loop BB32_34 Depth 3
                                        ;         Child Loop BB32_37 Depth 4
        .loc    212 57 12 is_stmt 1             ; src/details/ArborX_DetailsHappyTreeFriends.hpp:57:12
        v_lshlrev_b64 v[15:16], 5, v[13:14]
        v_add_co_u32_e32 v21, vcc, v11, v15
        v_addc_co_u32_e32 v22, vcc, v12, v16, vcc
        flat_load_dwordx4 v[15:18], v[21:22] offset:8
        flat_load_dwordx2 v[19:20], v[21:22] offset:24
.Ltmp7516:
        .loc    183 99 18                       ; src/details/ArborX_DetailsAlgorithms.hpp:99:18
        s_waitcnt vmcnt(0) lgkmcnt(0)
        v_cmp_nlt_f32_e32 vcc, v8, v15
        s_and_saveexec_b64 s[10:11], vcc
        s_cbranch_execz BB32_9
; %bb.6:                                ;   in Loop: Header=BB32_5 Depth=2
.Ltmp7517:
        .loc    183 101 23                      ; src/details/ArborX_DetailsAlgorithms.hpp:101:23
        v_cmp_ngt_f32_e32 vcc, v8, v18
        v_mov_b32_e32 v15, v18
        s_and_saveexec_b64 s[12:13], vcc
.Ltmp7518:
; %bb.7:                                ;   in Loop: Header=BB32_5 Depth=2
        .loc    183 0 23 is_stmt 0              ; src/details/ArborX_DetailsAlgorithms.hpp:0:23
        v_mov_b32_e32 v15, v8
; %bb.8:                                ; %Flow64
                                        ;   in Loop: Header=BB32_5 Depth=2
        s_or_b64 exec, exec, s[12:13]
BB32_9:                                 ;   in Loop: Header=BB32_5 Depth=2
        s_or_b64 exec, exec, s[10:11]
        .loc    183 99 18 is_stmt 1             ; src/details/ArborX_DetailsAlgorithms.hpp:99:18
        v_cmp_nlt_f32_e32 vcc, v9, v16
        s_and_saveexec_b64 s[10:11], vcc
        s_cbranch_execz BB32_13
; %bb.10:                               ;   in Loop: Header=BB32_5 Depth=2
.Ltmp7519:
        .loc    183 101 23                      ; src/details/ArborX_DetailsAlgorithms.hpp:101:23
        v_cmp_gt_f32_e32 vcc, v9, v19
        v_mov_b32_e32 v16, v9
        s_and_saveexec_b64 s[12:13], vcc
.Ltmp7520:
; %bb.11:                               ;   in Loop: Header=BB32_5 Depth=2
        .loc    183 0 23 is_stmt 0              ; src/details/ArborX_DetailsAlgorithms.hpp:0:23
        v_mov_b32_e32 v16, v19
; %bb.12:                               ; %Flow61
                                        ;   in Loop: Header=BB32_5 Depth=2
        s_or_b64 exec, exec, s[12:13]
BB32_13:                                ;   in Loop: Header=BB32_5 Depth=2
        s_or_b64 exec, exec, s[10:11]
        .loc    183 99 18 is_stmt 1             ; src/details/ArborX_DetailsAlgorithms.hpp:99:18
        v_cmp_nlt_f32_e32 vcc, v10, v17
        s_and_saveexec_b64 s[10:11], vcc
        s_cbranch_execz BB32_17
; %bb.14:                               ;   in Loop: Header=BB32_5 Depth=2
.Ltmp7521:
        .loc    183 101 23                      ; src/details/ArborX_DetailsAlgorithms.hpp:101:23
        v_cmp_gt_f32_e32 vcc, v10, v20
        v_mov_b32_e32 v17, v10
        s_and_saveexec_b64 s[12:13], vcc
.Ltmp7522:
; %bb.15:                               ;   in Loop: Header=BB32_5 Depth=2
        .loc    183 0 23 is_stmt 0              ; src/details/ArborX_DetailsAlgorithms.hpp:0:23
        v_mov_b32_e32 v17, v20
; %bb.16:                               ; %Flow59
                                        ;   in Loop: Header=BB32_5 Depth=2
        s_or_b64 exec, exec, s[12:13]
BB32_17:                                ; %_ZNK6ArborX10IntersectsINS_6SphereEEclINS_3BoxEEEbRKT_.exit.i.i.i
                                        ;   in Loop: Header=BB32_5 Depth=2
        s_or_b64 exec, exec, s[10:11]
```

When returning by value the data is simply loaded as 
```assembly
        flat_load_dwordx4 v[15:18], v[21:22] offset:8
        flat_load_dwordx2 v[19:20], v[21:22] offset:24
```
while when returning by reference this is not the case.

Ideally, someone with more knowledge of assembly can make sense of what goes wrong here. This is really important as we are running about two times slower with the currently generated code. One way or the other, we need to address this.

@Rombur What are other things we can try here, or other people we can invite to look at this code?

### PR DESCRIPTION
**Pull request by aprokop**
Fix #553

### CODE REVIEW

**Comment by aprokop** [comment]
# Spock results (0fd9c4a vs aae301f)

## Construction
Parity
```
BM_construction<ArborX::BVH<HIP>>/1000/0/manual_time_median                                              -0.0051         -0.0052           603           599           604           601
BM_construction<ArborX::BVH<HIP>>/1000/1/manual_time_median                                              +0.0026         +0.0026           612           614           614           616
BM_construction<ArborX::BVH<HIP>>/10000/0/manual_time_median                                             +0.0124         +0.0132           706           715           708           717
BM_construction<ArborX::BVH<HIP>>/10000/1/manual_time_median                                             +0.0117         +0.0124           712           721           714           723
BM_construction<ArborX::BVH<HIP>>/100000/0/manual_time_median                                            +0.0009         +0.0010          1016          1017          1038          1039
BM_construction<ArborX::BVH<HIP>>/100000/1/manual_time_median                                            -0.0001         -0.0001          1032          1032          1054          1054
BM_construction<ArborX::BVH<HIP>>/1000000/0/manual_time_median                                           +0.0099         +0.0097          2300          2322          2312          2335
BM_construction<ArborX::BVH<HIP>>/1000000/1/manual_time_median                                           +0.0007         +0.0016          2332          2334          2341          2345
```

## Radius search
20-60% improvement
```
BM_radius_search<ArborX::BVH<HIP>>/1000/1000/10/1/0/1/3/manual_time_median                               -0.1344         -0.1288          1195          1034          1247          1086
BM_radius_search<ArborX::BVH<HIP>>/1000/1000/10/1/0/0/2/manual_time_median                               -0.2186         -0.2121          1707          1334          1760          1386
BM_radius_search<ArborX::BVH<HIP>>/1000/1000/10/0/0/1/3/manual_time_median                               -0.2300         -0.2151           764           588           817           641
BM_radius_search<ArborX::BVH<HIP>>/1000/1000/10/0/0/0/2/manual_time_median                               -0.3225         -0.3101          1282           869          1333           920
BM_radius_search<ArborX::BVH<HIP>>/10000/10000/10/1/0/1/3/manual_time_median                             -0.1419         -0.1371          1423          1221          1475          1273
BM_radius_search<ArborX::BVH<HIP>>/10000/10000/10/1/0/0/2/manual_time_median                             -0.2563         -0.2501          2064          1535          2117          1587
BM_radius_search<ArborX::BVH<HIP>>/10000/10000/10/0/0/1/3/manual_time_median                             -0.2564         -0.2435           984           732          1036           784
BM_radius_search<ArborX::BVH<HIP>>/10000/10000/10/0/0/0/2/manual_time_median                             -0.4515         -0.4401          2061          1131          2114          1183
BM_radius_search<ArborX::BVH<HIP>>/100000/100000/10/1/0/1/3/manual_time_median                           -0.2008         -0.1942          2066          1651          2137          1722
BM_radius_search<ArborX::BVH<HIP>>/100000/100000/10/1/0/0/2/manual_time_median                           -0.3343         -0.3062          2951          1964          3221          2235
BM_radius_search<ArborX::BVH<HIP>>/100000/100000/10/0/0/1/3/manual_time_median                           -0.2344         -0.2217          1390          1064          1462          1138
BM_radius_search<ArborX::BVH<HIP>>/100000/100000/10/0/0/0/2/manual_time_median                           -0.5180         -0.4840          3879          1870          4149          2141
BM_radius_search<ArborX::BVH<HIP>>/1000000/1000000/10/1/0/1/3/manual_time_median                         -0.3315         -0.3175          6602          4413          6889          4702
BM_radius_search<ArborX::BVH<HIP>>/1000000/1000000/10/1/0/0/2/manual_time_median                         -0.4983         -0.4915         20542         10307         20830         10592
BM_radius_search<ArborX::BVH<HIP>>/1000000/1000000/10/0/0/1/3/manual_time_median                         -0.2678         -0.2578          8232          6028          8547          6344
BM_radius_search<ArborX::BVH<HIP>>/1000000/1000000/10/0/0/0/2/manual_time_median                         -0.5042         -0.5001         37542         18614         37858         18927
BM_radius_callback_search<ArborX::BVH<HIP>>/1000/1000/10/1/0/1/3/manual_time_median                      -0.1027         -0.0936           732           657           780           707
BM_radius_callback_search<ArborX::BVH<HIP>>/1000/1000/10/1/0/0/2/manual_time_median                      -0.1961         -0.1868           984           791          1033           840
BM_radius_callback_search<ArborX::BVH<HIP>>/1000/1000/10/0/0/1/3/manual_time_median                      -0.3661         -0.3047           240           152           288           200
BM_radius_callback_search<ArborX::BVH<HIP>>/1000/1000/10/0/0/0/2/manual_time_median                      -0.4310         -0.3940           492           280           539           327
BM_radius_callback_search<ArborX::BVH<HIP>>/10000/10000/10/1/0/1/3/manual_time_median                    -0.1182         -0.1093           807           712           854           760
BM_radius_callback_search<ArborX::BVH<HIP>>/10000/10000/10/1/0/0/2/manual_time_median                    -0.2403         -0.2291          1136           863          1183           912
BM_radius_callback_search<ArborX::BVH<HIP>>/10000/10000/10/0/0/1/3/manual_time_median                    -0.3630         -0.3165           344           219           391           267
BM_radius_callback_search<ArborX::BVH<HIP>>/10000/10000/10/0/0/0/2/manual_time_median                    -0.5363         -0.5079           887           411           934           460
BM_radius_callback_search<ArborX::BVH<HIP>>/100000/100000/10/1/0/1/3/manual_time_median                  -0.1936         -0.1829          1127           909          1191           973
BM_radius_callback_search<ArborX::BVH<HIP>>/100000/100000/10/1/0/0/2/manual_time_median                  -0.3250         -0.3118          1540          1039          1604          1104
BM_radius_callback_search<ArborX::BVH<HIP>>/100000/100000/10/0/0/1/3/manual_time_median                  -0.3153         -0.2787           494           338           558           402
BM_radius_callback_search<ArborX::BVH<HIP>>/100000/100000/10/0/0/0/2/manual_time_median                  -0.5895         -0.5684          1717           705          1780           768
BM_radius_callback_search<ArborX::BVH<HIP>>/1000000/1000000/10/1/0/1/3/manual_time_median                -0.3296         -0.3151          3456          2317          3612          2474
BM_radius_callback_search<ArborX::BVH<HIP>>/1000000/1000000/10/1/0/0/2/manual_time_median                -0.5120         -0.5042         10204          4979         10359          5136
BM_radius_callback_search<ArborX::BVH<HIP>>/1000000/1000000/10/0/0/1/3/manual_time_median                -0.2985         -0.2838          3472          2436          3657          2619
BM_radius_callback_search<ArborX::BVH<HIP>>/1000000/1000000/10/0/0/0/2/manual_time_median                -0.5782         -0.5721         17842          7526         18017          7710
```

## knn search
0-25% improvement
```
BM_knn_search<ArborX::BVH<HIP>>/1000/1000/10/1/1/3/manual_time_median                                    -0.0693         -0.0674          1919          1786          1971          1838
BM_knn_search<ArborX::BVH<HIP>>/1000/1000/10/1/0/2/manual_time_median                                    -0.0962         -0.0943          2051          1854          2104          1905
BM_knn_search<ArborX::BVH<HIP>>/1000/1000/10/0/1/3/manual_time_median                                    -0.1158         -0.1118          1517          1342          1570          1394
BM_knn_search<ArborX::BVH<HIP>>/1000/1000/10/0/0/2/manual_time_median                                    -0.1363         -0.1322          1628          1406          1679          1457
BM_knn_search<ArborX::BVH<HIP>>/10000/10000/10/1/1/3/manual_time_median                                  -0.0990         -0.0972          2738          2467          2790          2519
BM_knn_search<ArborX::BVH<HIP>>/10000/10000/10/1/0/2/manual_time_median                                  -0.0981         -0.0959          2269          2046          2321          2099
BM_knn_search<ArborX::BVH<HIP>>/10000/10000/10/0/1/3/manual_time_median                                  -0.1968         -0.1927          2457          1973          2509          2026
BM_knn_search<ArborX::BVH<HIP>>/10000/10000/10/0/0/2/manual_time_median                                  -0.2054         -0.2002          2046          1626          2099          1679
BM_knn_search<ArborX::BVH<HIP>>/100000/100000/10/1/1/3/manual_time_median                                -0.1015         -0.0975          5686          5109          5904          5328
BM_knn_search<ArborX::BVH<HIP>>/100000/100000/10/1/0/2/manual_time_median                                -0.0665         -0.0624          3435          3207          3655          3427
BM_knn_search<ArborX::BVH<HIP>>/100000/100000/10/0/1/3/manual_time_median                                -0.1857         -0.1789          5865          4776          6083          4995
BM_knn_search<ArborX::BVH<HIP>>/100000/100000/10/0/0/2/manual_time_median                                -0.2039         -0.1924          3690          2937          3909          3157
BM_knn_search<ArborX::BVH<HIP>>/1000000/1000000/10/1/1/3/manual_time_median                              -0.0884         -0.0876         38133         34763         38477         35107
BM_knn_search<ArborX::BVH<HIP>>/1000000/1000000/10/1/0/2/manual_time_median                              -0.0718         -0.0708         22017         20436         22366         20782
BM_knn_search<ArborX::BVH<HIP>>/1000000/1000000/10/0/1/3/manual_time_median                              -0.1389         -0.1383         76996         66301         77259         66574
BM_knn_search<ArborX::BVH<HIP>>/1000000/1000000/10/0/0/2/manual_time_median                              -0.1197         -0.1163         35387         31150         35669         31520
BM_knn_callback_search<ArborX::BVH<HIP>>/1000/1000/10/1/1/3/manual_time_median                           -0.1057         -0.1019          1443          1290          1490          1338
BM_knn_callback_search<ArborX::BVH<HIP>>/1000/1000/10/1/0/2/manual_time_median                           -0.1309         -0.1270          1564          1360          1612          1408
BM_knn_callback_search<ArborX::BVH<HIP>>/1000/1000/10/0/1/3/manual_time_median                           -0.1622         -0.1553          1101           922          1149           970
BM_knn_callback_search<ArborX::BVH<HIP>>/1000/1000/10/0/0/2/manual_time_median                           -0.1807         -0.1738          1231          1008          1279          1057
BM_knn_callback_search<ArborX::BVH<HIP>>/10000/10000/10/1/1/3/manual_time_median                         -0.1232         -0.1203          2314          2029          2361          2077
BM_knn_callback_search<ArborX::BVH<HIP>>/10000/10000/10/1/0/2/manual_time_median                         -0.1271         -0.1236          1839          1605          1886          1653
BM_knn_callback_search<ArborX::BVH<HIP>>/10000/10000/10/0/1/3/manual_time_median                         -0.2422         -0.2365          2049          1553          2097          1601
BM_knn_callback_search<ArborX::BVH<HIP>>/10000/10000/10/0/0/2/manual_time_median                         -0.2648         -0.2570          1636          1203          1683          1251
BM_knn_callback_search<ArborX::BVH<HIP>>/100000/100000/10/1/1/3/manual_time_median                       -0.1146         -0.1111          5076          4494          5231          4650
BM_knn_callback_search<ArborX::BVH<HIP>>/100000/100000/10/1/0/2/manual_time_median                       -0.0838         -0.0794          2838          2601          2994          2756
BM_knn_callback_search<ArborX::BVH<HIP>>/100000/100000/10/0/1/3/manual_time_median                       -0.2055         -0.1995          5292          4204          5448          4361
BM_knn_callback_search<ArborX::BVH<HIP>>/100000/100000/10/0/0/2/manual_time_median                       -0.2444         -0.2323          3111          2351          3268          2509
BM_knn_callback_search<ArborX::BVH<HIP>>/1000000/1000000/10/1/1/3/manual_time_median                     -0.1003         -0.0996         35509         31947         35746         32185
BM_knn_callback_search<ArborX::BVH<HIP>>/1000000/1000000/10/1/0/2/manual_time_median                     -0.0844         -0.0833         19635         17979         19877         18221
BM_knn_callback_search<ArborX::BVH<HIP>>/1000000/1000000/10/0/1/3/manual_time_median                     -0.1434         -0.1430         75261         64466         75514         64718
BM_knn_callback_search<ArborX::BVH<HIP>>/1000000/1000000/10/0/0/2/manual_time_median                     -0.1331         -0.1321         33581         29110         33847         29374
```

**Comment by aprokop** [comment]
Overall, I think it's fine to be conservative and enable both checks making sure it just affects device code compiled with hipcc.

**Comment by aprokop** [comment]
DBSCAN with HACC data:
```
spock25 $ ./ArborX_DBSCAN_master_0fd9c4a  --binary --eps 0.042 --filename ~/hacc.arborx --print-dbscan-timers --impl fdbscan
ArborX version    : 1.2 (dev)
ArborX hash       : 0fd9c4a
eps               : 0.042000
minpts            : 2
cluster min size  : 1
filename          : /ccs/home/aprokop/hacc.arborx [binary, max_pts = -1]
filename [labels] :  [binary]
implementation    : fdbscan
samples           : -1
verify            : false
print timers      : true
output centers    : false
Reading in "/ccs/home/aprokop/hacc.arborx" in binary mode...done
Read in 36902719 3D points
-- construction     :      0.038
-- query+cluster    :      0.998
-- postprocess      :      0.010
total time          :      1.067

#clusters       : 2763469
#cluster points : 23609948 [63.98%]
```
```
spock25 $ ./ArborX_DBSCAN_hip_aae301f  --binary --eps 0.042 --filename ~/hacc.arborx --print-dbscan-timers --impl fdbscan
ArborX version    : 1.2 (dev)
ArborX hash       : aae301f
eps               : 0.042000
minpts            : 2
cluster min size  : 1
filename          : /ccs/home/aprokop/hacc.arborx [binary, max_pts = -1]
filename [labels] :  [binary]
implementation    : fdbscan
samples           : -1
verify            : false
print timers      : true
output centers    : false
Reading in "/ccs/home/aprokop/hacc.arborx" in binary mode...done
Read in 36902719 3D points
-- construction     :      0.038
-- query+cluster    :      0.523
-- postprocess      :      0.012
total time          :      0.595

#clusters       : 2763469
#cluster points : 23609948 [63.98%]
```

File: src/details/ArborX_DetailsHappyTreeFriends.hpp
Comment (masterleinad) [inline src/details/ArborX_DetailsHappyTreeFriends.hpp:49]:
```suggestion
  // FIXME_HIP
  // The version of the function returning const ref runs significantly slower
```

**Review by masterleinad — COMMENTED** [review]
This is fine with me but we should have a FIXME note so that we don't forget to remove it eventually when the HIP compiler team has improved the behavior.

**Comment by aprokop** [comment]
The test run on Ascent shows no affect on Cuda as expected.

**Review by masterleinad — APPROVED** [review]

**Comment by dalg24** [comment]
Please check whether this improves performance when using Clang+CUDA before we merge.

File: src/details/ArborX_DetailsHappyTreeFriends.hpp
Comment (dalg24) [inline src/details/ArborX_DetailsHappyTreeFriends.hpp:48]: Why is it not just `#ifdef __HIP_DEVICE_COMPILE__`?

**Comment by aprokop** [comment]
> Please check whether this improves performance when using Clang+CUDA before we merge.

As `__HIPCC__` and `__HIP_DEVICE_COMPILE__` are not defined for Clang, it should not affect it.

File: src/details/ArborX_DetailsHappyTreeFriends.hpp
Comment (aprokop) [reply src/details/ArborX_DetailsHappyTreeFriends.hpp:48]: https://rocmdocs.amd.com/en/latest/Programming_Guides/HIP-porting-guide.html#compiler-defines-summary

My original code was
`#if defined(__HIP__) && defined(__HIP_DEVICE_COMPILE__)`
as I conservatively wanted to only enable it for the device code and HIP-Clang compiler. `__HIP__` was only defined for that.

@Rombur asked to change it to `__HIPCC__`. But that is now defined for both HIP-Clang and nvcc. So, I guess, now it's not necessary in the presence of `__HIP_DEVICE_COMPILE__`. I can remove `__HIPCC__` or reintroduce `__HIP__`.

Comment (dalg24) [reply src/details/ArborX_DetailsHappyTreeFriends.hpp:48]: We are never compiling with `hipcc` to target NVIDIA hardware (not allowed by Kokkos) so this is pointless

**Comment by dalg24** [comment]
> > Please check whether this improves performance when using Clang+CUDA before we merge.
> 
> As `__HIPCC__` and `__HIP_DEVICE_COMPILE__` are not defined for Clang, it will not affect it.

We need to check if that optimization is relevant for Clang+CUDA.  We have discussed that at the time we discovered the issue.

**Comment by aprokop** [comment]
> We need to check if that optimization is relevant for Clang+CUDA. We have discussed that at the time we discovered the issue.

I compiled Kokkos on Summit using clang/12.0.0. and cuda/10.1.243 (as clang/12 only officially supports cuda up to 10.1). I tried building ArborX with this combination, but was unsuccessful:
```
ptxas /tmp/dbscan-a808a6.s, line 77662; error   : Instruction 'shfl' without '.sync' is not supported on .target sm_70 and higher from PTX ISA version 6.4
ptxas fatal   : Ptx assembly aborted due to errors
clang-12: error: ptxas command failed with exit code 255 (use -v to see invocation)
clang version 12.0.0 (https://github.com/llvm/llvm-project.git d28af7c654d8db0b68c175db5ce212d74fb5e9bc)
Target: powerpc64le-unknown-linux-gnu
Thread model: posix
InstalledDir: /sw/summit/llvm/12.0.0/12.0.0-0/bin
clang-12: note: diagnostic msg:
```
I switched to cuda/10.2.89. That was not successful either.
Trying to compile Kokkos with llvm/10.0.1 and cuda/10.1.243 was not successful either:
```
In file included from <built-in>:1:
In file included from /autofs/nccs-svm1_sw/summit/llvm/10.0.1/10.0.1-0/lib/clang/10.0.1/include/__clang_cuda_runtime_wrapper.h:36:
In file included from /usr/lib/gcc/ppc64le-redhat-linux/8/../../../../include/c++/8/cmath:45:
In file included from /usr/include/math.h:290:
/usr/include/bits/mathcalls.h:107:13: error: exception specification in declaration does not match previous declaration
__MATHCALL (log10,, (_Mdouble_ __x));
```

In short, if you have an environment setup, please test it out. Otherwise, please provide detailed instructions on how to build it in this mode.

File: src/details/ArborX_DetailsHappyTreeFriends.hpp
Comment (dalg24) [reply src/details/ArborX_DetailsHappyTreeFriends.hpp:48]: @Rombur please advise on how to best guard the code here

**Comment by dalg24** [comment]
Per https://github.com/arborx/ArborX/issues/553#issuecomment-985888894, I am basically OK to merge this once we settle how to guard the code.

File: src/details/ArborX_DetailsHappyTreeFriends.hpp
Comment (dalg24) [inline src/details/ArborX_DetailsHappyTreeFriends.hpp:60]:
```suggestion
// FIXME_HIP See https://github.com/arborx/ArborX/issues/553
#ifdef __HIP_DEVICE_COMPILE__
  static KOKKOS_FUNCTION auto getBoundingVolume(BVH const &bvh, int i)
#else
```

File: src/details/ArborX_DetailsHappyTreeFriends.hpp
Comment (Rombur) [reply src/details/ArborX_DetailsHappyTreeFriends.hpp:48]: Agree `__HIPCC__` is useless and so would be `__HIP__`

**Comment by aprokop** [comment]
OK, let me just quickly test it on Spock to make sure everything is as expected.

**Review by dalg24 — APPROVED** [review]

**Comment by aprokop** [comment]
> OK, let me just quickly test it on Spock to make sure everything is as expected.

Everything check out.

**Comment by dalg24** [comment]
Failure are warnings emitted during build of the SYCL image.  This is ready to merge.

**Pull request merged**
